Implemented user friends feature

MalikZu opened the pull request.

MalikZu committed:
feat(Users Module): added friendship relation

Taanviir commented on backend/users/views.py:
```diff
@@ -243,3 +243,38 @@ def change_password_view(request):
+
+
+##############
+##### Frineds Methods
```
typo

Taanviir commented on backend/users/views.py:
```diff
@@ -243,3 +243,38 @@ def change_password_view(request):
+
+@jwt_login_required
+def add_friend(request, user_id):
+    friend = get_object_or_404(CustomUser, id=user_id)
```
what happens in the case of 404

MalikZu replied:
don't we have default 404 page? This function on django as I understand helps with detecting the error and by default it will redirect to 404 page if it fails

Taanviir replied:
we do teh default 404 only works with django running in production mode

Taanviir commented on backend/users/views.py:
```diff
@@ -243,3 +243,38 @@ def change_password_view(request):
+@jwt_login_required
+def friend_list(request):
+    friends = request.user.get_friends()
+    return render(request, "users/friend_list.html", {"friends": friends})
```
this html file is not in this PR

MalikZu replied:
exactly, it depends on frontend, i assumed these two pages will be created by you

Taanviir replied:
alright that makes sense

Taanviir commented on backend/users/views.py:
```diff
@@ -243,3 +243,38 @@ def change_password_view(request):
+        friend__last_activity__gte=timezone.now() - timezone.timedelta(minutes=5)
+    )
+    return render(
+        request, "users/online_friends.html", {"online_friends": online_friends}
```
and this html file too

Taanviir commented on backend/users/models.py:
```diff
@@ -16,3 +17,47 @@ class CustomUser(AbstractUser):
     is_authenticated = models.BooleanField(default=False)
     is_42 = models.BooleanField(default=False)
+    last_activity = models.DateTimeField(default=timezone.now)
+    friends = models.ManyToManyField(
```
could you explain how this relationship works so i have a better idea of this

MalikZu replied:
so every User will have `friends` list, but to define the relation between two friends i am using "Friendship" class. So every User will have list of Friendships that will be defined as Friendship object/model

Taanviir replied:
that makes sense, so this way one user can have multiple friendship objects and then refer to their friends

MalikZu committed:
chore: add test for frienships model

MalikZu committed:
feat(user-management): add accept/reject Functionality for frienship request

MalikZu committed:
chore: update getters to support approval of friendship

MalikZu committed:
chore: update tests for friendship model

MalikZu committed:
feat(user-management): add matches history with wins and losses

MalikZu committed:
feat(user-management): add views

MalikZu committed:
chore: update tests

Taanviir commented on backend/templates/users/add_friend.html:
```diff
@@ -0,0 +1,13 @@
+    <form method="POST">
+        {% csrf_token %}
+        <input type="text" name="username" placeholder="Enter username or email" required>
+        <button type="submit" class="btn btn-primary">Send Friend Request</button>
```
When a friend request is sent from here to the other user, the other user has no way of knowing they've received a request until they basically click on the "FRIENDS LIST" menu button as in they have to send a request to the backend

MalikZu replied:
we can add a badge on the Friend list when there is a pending requests.

MalikZu replied:
Done, please check.

Taanviir commented on backend/templates/users/friend_list.html:
```diff
@@ -0,0 +1,50 @@
+        {% for request in pending_requests %}
+        <li>
+            {{ request.username }} wants to be your friend.
+             <a href="{% url 'accept_friend' request.username %}">Accept</a> |
```
Clicking this causes a 404 error

Taanviir replied:
![image](https://github.com/user-attachments/assets/36cdb76c-7b95-49d8-bd94-0f2436174d9b)

MalikZu replied:
now I changed the whole 404 page thing, to error modal, please try it again. The errors themselves might be related to logic, but now it makes sense i think it term of view and handling

Taanviir commented on backend/templates/users/friend_list.html:
```diff
@@ -0,0 +1,50 @@
+        <li>
+            {{ request.username }} wants to be your friend.
+             <a href="{% url 'accept_friend' request.username %}">Accept</a> |
+            <a href="{% url 'reject_friend' request.username %}">Reject</a>
```
same behavior with this

Taanviir commented on backend/users/views.py:
```diff
@@ -243,3 +244,98 @@ def change_password_view(request):
+@jwt_login_required
+def add_friend(request, username):
+    friend = get_object_or_404(CustomUser, username=username)
+    request.user.add_friend(friend)
```
what happens if this request fails, like if the friend does not exist?

Taanviir replied:
![image](https://github.com/user-attachments/assets/3242f638-4580-49f0-b96d-40317824fc3e)

MalikZu replied:
this is in Debug mode true; if it false it should go to 404 page.

Taanviir replied:
i'll find a way for the user to get notified in the page itself instead

MalikZu committed:
feat(user-management): badge pending requests indicator

MalikZu committed:
feat: modal for error messages related to friendship

MalikZu committed:
chore(user-management): revamp errors related to friendship to use error modal not 404 pages

MalikZu committed:
fix(user-management): user doesn't have to have auth

MalikZu committed:
chore: don't allow same user to add himself as a friend

Taanviir commented on backend/templates/users/friend_list.html:
```diff
@@ -0,0 +1,80 @@
+</div>
+
+<!-- JavaScript to handle the modal -->
+<script>
+    document.addEventListener('DOMContentLoaded', function() {
+        const errorMessage = "{{ error_message|escapejs }}";
+        if (errorMessage) {
+            document.getElementById('errorMessage').innerText = errorMessage;
+            const errorModal = new bootstrap.Modal(document.getElementById('errorModal'));
+            errorModal.show();
+        }
+    });
+</script>
+<!-- Modal for error messages -->
+<div class="modal fade" id="errorModal" tabindex="-1" aria-labelledby="errorModalLabel" aria-hidden="true" data-bs-theme="dark">
+    <div class="modal-dialog">
+        <div class="modal-content">
+            <div class="modal-header">
+                <h5 class="modal-title" id="errorModalLabel">Error</h5>
+                <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
+            </div>
+            <div class="modal-body">
+                <p id="errorMessage" ></p>
+            </div>
+            <div class="modal-footer">
+                <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
+            </div>
+        </div>
+    </div>
+</div>
```
This modal causes a reload to happen. I believe it is because in views.py you are rendering the html however instead of using spa routing functions like [getPage()](https://github.com/Taanviir/ft_transcendence/blob/main/backend/static/scripts/main.js#L72) defined in main.js, you are rendering the html directly

MalikZu replied:
I am trying to avoid using js functionalities in the front end and rather use the backend to handle error and allow functionality, specially that I think it is cleaner this way. But we can go to this later after we are sure that it works well.

Taanviir commented on backend/users/views.py:
```diff
@@ -243,3 +244,151 @@ def change_password_view(request):
+
+    if not friend:
+        error_message = "User not found."
+    elif friend == request.user:
+        error_message = "You cannot add yourself as a friend."
```
This case is not working, I am able to add myself as friend.

Taanviir replied:
![image](https://github.com/user-attachments/assets/b41d84f7-884a-48ef-9681-4575621b821d)

MalikZu replied:
check now?

Taanviir replied:
works

Taanviir commented on backend/users/views.py:
```diff
@@ -243,3 +244,151 @@ def change_password_view(request):
+
+
+@jwt_login_required
+def remove_friend(request, username):
```
Is there a button to remove friend on the frontend?

MalikZu replied:
done

MalikZu committed:
feat: remove friend button on FE

MalikZu committed:
feat: create friendship for each party

MalikZu committed:
Merge branch 'main' into user-managment-module

Commit:
Merge branch 'main' into user-managment-module

MalikZu committed:
chore: refactor models

MalikZu committed:
fix: online friends

MalikZu committed:
refactor: placeholder of adding friend

MalikZu committed:
feat: email is unique now.

MalikZu committed:
feat: add sucess message, remove friend without reload

MalikZu committed:
Merge branch 'user-managment-module' of https://github.com/Taanviir/ft_transcendence into user-managment-module

MalikZu committed:
chore: missing modal

Taanviir committed:
feat: friends functionality doesnt refresh now

- Backend: moved friend related functions into one file
- Frontend: friend.js contains functions that handle adding, removing,
  accepting and rejecting friends
- merged add_friend and friend_list into one
- updated friends django functions to not render anything

Taanviir committed:
feat: updated friends.html design

MalikZu commented on backend/users/friends.py:
Why not have it under views folder? since now it is a separate file, at the end of the day this falls under views.

Taanviir replied:
good idea, i will do this now

MalikZu reviewed (commented):
looks good to me

MalikZu commented:
@Taanviir Please resolve conflicts (since they are related to nav_menu template) and merge,

Taanviir committed:
Merge branch 'main' into user-managment-module

Taanviir committed:
feat(backend): moved api_42, views, and friends to views/

Taanviir committed:
chore: removed add_friend.html
